what4/src/What4/Expr/WeightedSum.hs

```diff
@@ -574,36 +621,43 @@ extractCommon (WeightedSum xm xc sr) (WeightedSum ym yc _) = (z, x', y')
 
 -- | Returns true if the product is trivial (contains no terms).
 nullProd :: SemiRingProduct f sr -> Bool
-nullProd pd = Map.null (_prodMap pd)
+nullProd pd = AM.null (_prodMap pd)
 
 -- | If the product consists of exactly on term, return it.
 asProdVar :: SemiRingProduct f sr -> Maybe (f (SR.SemiRingBase sr))
 asProdVar pd
-  | [(WrapF x, SR.occ_count sr -> 1)] <- Map.toList (_prodMap pd) = Just x
+  | [(WrapF x, SR.occ_count sr -> 1)] <- AM.toList (_prodMap pd) = Just x
   | otherwise = Nothing
  where
  sr = prodRepr pd
 
+prodAbsValue :: OrdF f => SemiRingProduct f sr -> AD.AbstractValue (SR.SemiRingBase sr)
+prodAbsValue pd =
+  fromSRAbsValue $
+  case AM.annotation (_prodMap pd) of
+    Nothing             -> abstractScalar (prodRepr pd) (SR.one (prodRepr pd))
+    Just (ProdNote _ v) -> v
+
 -- | Returns true if the product contains at least on occurence of the given term.
 prodContains :: OrdF f => SemiRingProduct f sr -> f (SR.SemiRingBase sr) -> Bool
-prodContains pd x = Map.member (WrapF x) (_prodMap pd)
+prodContains pd x = isJust $ AM.lookup (WrapF x) (_prodMap pd)
 
 -- | Produce a product map from a raw map of terms to occurences.
 --   PRECONDITION: the occurence value for each term should be non-zero.
-mkProd :: HashableF f => SR.SemiRingRepr sr -> Map (WrapF f sr) (SR.Occurrence sr) -> SemiRingProduct f sr
-mkProd sr m = SemiRingProduct m (computeProdHash sr m) sr
+mkProd :: HashableF f => SR.SemiRingRepr sr -> ProdMap f sr -> SemiRingProduct f sr
+mkProd sr m = SemiRingProduct m sr
 
 -- | Produce a product representing the single given term.
 prodVar :: Tm f => SR.SemiRingRepr sr -> f (SR.SemiRingBase sr) -> SemiRingProduct f sr
-prodVar sr x = mkProd sr (Map.singleton (WrapF x) (SR.occ_one sr))
+prodVar sr x = mkProd sr (singletonProdMap sr (SR.occ_one sr) x)
 
 -- | Multiply two products, collecting terms and adding occurences.
 prodMul :: Tm f => SemiRingProduct f sr -> SemiRingProduct f sr -> SemiRingProduct f sr
 prodMul x y = mkProd sr m
   where
   sr = prodRepr x
-  mergeCommon _ a b = Just (SR.occ_add sr a b)
-  m = Map.mergeWithKey mergeCommon id id (_prodMap x) (_prodMap y)
+  mergeCommon _ (u,a) (v,b) = Just (u <> v, SR.occ_add sr a b)
```

Comment:
Oh... pretty sure this isn't right, actually.  This will get the wrong hash value

Reply:
Yeah, you're going to want to call `mkProdNote` there, just like `mkNote` is used in the implementation of function `add`